server/svix-server/src/worker.rs

```diff
@@ -43,6 +46,122 @@ const USER_AGENT: &str = concat!("Svix-Webhooks/", env!("CARGO_PKG_VERSION"));
 /// Send the MessageAttemptFailingEvent after exceeding this number of failed attempts
 const OP_WEBHOOKS_SEND_FAILING_EVENT_AFTER: usize = 4;
 
+/// A simple enum noting whether to disable or not to disable an endpoint. Returned from the
+/// [`process_failure_cache`] function which is to be called after all retry events are exhausted.
+enum EndpointDisableStatus {
+    DoNotDisable,
+    DoDisable { first_failure_at: DateTimeUtc },
+}
+
+/// The first_failure_at time is only stored in Postgres after the endpoint has been disabled.
+/// Otherwise, it is stored in the cache with an expiration.
+#[derive(Deserialize, Serialize)]
+pub struct FailureCacheValue {
+    pub first_failure_at: DateTimeUtc,
+}
+
+/// The key is simply a formatted [`String`] containing the [`ApplicationId`] and [`EndpointId`]
+/// of the endpoint which exhausted the retry schedule.
+pub struct FailureCacheKey {
+    key: String,
+}
+
+impl FailureCacheKey {
+    pub fn new(app_id: &ApplicationId, endp_id: &EndpointId) -> FailureCacheKey {
+        FailureCacheKey {
+            key: format!("_{}_{}", app_id, endp_id),
+        }
+    }
+}
+
+impl AsRef<str> for FailureCacheKey {
+    fn as_ref(&self) -> &str {
+        &self.key
+    }
+}
+
+impl CacheKey for FailureCacheKey {
+    const PREFIX_CACHE: &'static str = "SVIX_FAILURE_CACHE";
+}
+
+impl CacheValue for FailureCacheValue {
+    type Key = FailureCacheKey;
+}
+
+/// Called upon the successful dispatch of an endpoint. Simply clears the cache of a
+/// [`FailureCacheKey`]/[`FailureCacheValue`] pair associated with a given endpoint. This is such
+/// that an endpoint that was previously not responding is not disabled after responding again.
+///
+/// If the key value pair does not already exist in the cache, indicating that the endpoint never
+/// stopped responding, no operation is performed.
+async fn process_success_cache(
+    cache: &Cache,
+    app_id: &ApplicationId,
+    endp_id: &EndpointId,
+) -> Result<()> {
+    let key = FailureCacheKey::new(app_id, endp_id);
+
+    // XXX: Make sure to test deleting a value that does not exist doesn't cause an error response
```

Comment:
Code is no place for TODO items. These kind of comments are useful for stuff specific to people changing the code, but this is not the case here.